### PR TITLE
Fixed logic of SEMI in clustering

### PR DESCRIPTION
**Pull request by Vitaly-Protasov**
Directly based on the https://github.com/cqfn/veniq/pull/18

### CODE REVIEW

File: veniq/baselines/semi/clustering.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    first_statement_ = list(dict_file.keys())[0]
+    last_statement_ = list(dict_file.keys())[-1]
+    first_statement = first_statement_ if isinstance(first_statement_, int) else first_statement_.line
+    last_statement = last_statement_ if isinstance(last_statement_, int) else last_statement_.line
+    method_length = last_statement - first_statement
+
+    for step in range(1, method_length + 1):
+        clusters = _process_statement(dict_file, step)
+        opportunities += clusters
+    unique_oppo = [list(oppo) for oppo in set(map(tuple, opportunities))]
```
Comment (acheshkov) [inline veniq/baselines/semi/clustering.py:55]: 1. why do you need a `set` operator? also `tuple` seems no need?
2. Line 55 looks redundant. Can we return `opportunities` instead?

Comment (Vitaly-Protasov) [reply veniq/baselines/semi/clustering.py:55]: I think it is clear now from our talk

**Comment by KatGarmash** [comment]
@Vitaly-Protasov which part of the SEMI pipeline does this PR implement?

File: veniq/baselines/semi/clustering.py [hunk 2]
```diff
@@ -17,36 +17,71 @@ def check_is_common(
...
         stat_1_line = stat_1 if isinstance(stat_1, int) else stat_1.line
         for stat_2 in list_statements[:stat_1_line + step]:
             stat_2_line = stat_2 if isinstance(stat_2, int) else stat_2.line
-            if stat_1_line < stat_2_line and check_is_common(dict_file, stat_1, stat_2):
-                if len(clusters) != 0 and is_in_range(stat_1_line, clusters[-1]):
-                    if not is_in_range(stat_2_line, clusters[-1]):
+            if stat_1_line < stat_2_line and _check_is_common(dict_file, stat_1, stat_2):
+                if len(clusters) != 0 and _is_in_range(stat_1_line, clusters[-1]):
+                    if not _is_in_range(stat_2_line, clusters[-1]):
                         clusters[-1][1] = stat_2_line
```
Comment (KatGarmash) [inline veniq/baselines/semi/clustering.py:37]: what does this line do?

**Review by KatGarmash — CHANGES_REQUESTED** [review]

File: veniq/baselines/semi/clustering.py [hunk 3]
```diff
@@ -66,17 +101,10 @@ def _print_clusters(
     method_semantic = extract_method_statements_semantic(method_ast)
     reporcessed_dict = _reprocess_dict(method_semantic)
```
[the remark this answers is not in the document]
Comment (Vitaly-Protasov) [reply veniq/baselines/semi/clustering.py:102]: +

File: veniq/baselines/semi/clustering.py [hunk 4]
```diff
@@ -66,17 +101,10 @@ def _print_clusters(
     method_semantic = extract_method_statements_semantic(method_ast)
```
[the remark this answers is not in the document]
Comment (Vitaly-Protasov) [reply veniq/baselines/semi/clustering.py:101]: so what? yes it works like that

File: veniq/baselines/semi/clustering.py [hunk 5]
```diff
@@ -17,36 +17,71 @@ def check_is_common(
...
+    return clusters_by_nodes
+
+
+def SEMI_ASTNodes(dict_file: Dict[ASTNode, List[str]]) -> List[List[ASTNode]]:
+    clusters: List[List[int]] = SEMI(dict_file)
+    lines_to_node_dict = _get_lines_to_node_dict(dict_file)
+    return _transform_clusters(clusters, lines_to_node_dict)
 
 
 def _reprocess_dict(method_semantic: Dict) -> Dict[ASTNode, List[str]]:
```
[the remark this answers is not in the document]
Comment (Vitaly-Protasov) [reply veniq/baselines/semi/clustering.py:87]: +++

File: veniq/baselines/semi/clustering.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Vitaly-Protasov) [reply veniq/baselines/semi/clustering.py:102]: +++

File: veniq/baselines/semi/clustering.py [hunk 6]
```diff
@@ -17,36 +17,71 @@ def check_is_common(
...
 
 
-def SEMI_beta(dict_file: Dict, method_len: int) -> Dict[int, List[List[int]]]:
-    algo_step = {}
-    statements = list(dict_file.keys())
-    for step in range(1, method_len + 1):
-        clusters = process_statement(dict_file, statements, step)
-        algo_step[step] = clusters
-    return algo_step
+def SEMI(dict_file: Dict[ASTNode, List[str]]) -> List[List[int]]:
```
[the remark this answers is not in the document]
Comment (Vitaly-Protasov) [reply veniq/baselines/semi/clustering.py:43]: ++++

File: veniq/baselines/semi/clustering.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (KatGarmash) [reply veniq/baselines/semi/clustering.py:101]: why do u include an unused argument?

File: veniq/baselines/semi/clustering.py [hunk 7]
```diff
@@ -17,36 +17,71 @@ def check_is_common(
...
+    opportunities = []
+
+    first_statement_ = list(dict_file.keys())[0]
+    last_statement_ = list(dict_file.keys())[-1]
+    first_statement = first_statement_ if isinstance(first_statement_, int) else first_statement_.line
+    last_statement = last_statement_ if isinstance(last_statement_, int) else last_statement_.line
+    method_length = last_statement - first_statement
```
[the remark this answers is not in the document]
Comment (Vitaly-Protasov) [reply veniq/baselines/semi/clustering.py:50]: done

File: veniq/baselines/semi/clustering.py [hunk 8]
```diff
@@ -17,36 +17,71 @@ def check_is_common(
     return len(list(duplicates)) >= 1
 
 
-def is_in_range(elem: int, values: List[int]) -> bool:
+def _is_in_range(elem: int, values: List[int]) -> bool:
     return elem >= values[0] and elem <= values[1]
 
 
-def process_statement(
+def _process_statement(
```
[the remark this answers is not in the document]
Comment (Vitaly-Protasov) [reply veniq/baselines/semi/clustering.py:24]: done

File: veniq/baselines/semi/clustering.py [hunk 9]
```diff
@@ -7,7 +7,7 @@
 from veniq.baselines.semi.common_cli import common_cli
 
 
-def check_is_common(
+def _check_is_common(
```
[the remark this answers is not in the document]
Comment (Vitaly-Protasov) [reply veniq/baselines/semi/clustering.py:10]: done

File: veniq/baselines/semi/clustering.py [hunk 10]
```diff
@@ -17,36 +17,71 @@ def check_is_common(
     return len(list(duplicates)) >= 1
 
 
-def is_in_range(elem: int, values: List[int]) -> bool:
+def _is_in_range(elem: int, values: List[int]) -> bool:
```
[the remark this answers is not in the document]
Comment (Vitaly-Protasov) [reply veniq/baselines/semi/clustering.py:20]: done

File: veniq/baselines/semi/clustering.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (KatGarmash) [reply veniq/baselines/semi/clustering.py:124]: ?

**Review by KatGarmash — APPROVED** [review]

**Pull request merged**